### ISSUE TITLE
Fused-compiled training plan drops gradients for 26/29 trainable params in standard Transformer (not just EmbeddingLayer)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1331]
## Summary

The fused-compiled training path (`CompiledTapeTrainingStep.TryStepWithFusedOptimizer` → `AiDotNet.Tensors.CompiledTrainingPlan.Step`) silently drops dL/dW for most trainable parameters in a standard Transformer encoder. Only 3 of 29 params (the additive biases on the final ~3 layers) receive non-zero gradients. The remaining 26 params — including ALL 2D weight matrices, the embedding table, and most LayerNorm gain/beta — are bit-identical after a `Train()` call.

This is a **broader fused-path regression** than #1328 (which was scoped to `EmbeddingLayer`). The engine-level fix in `AiDotNet.Tensors` #354 (TensorEmbeddingLookup → GraphMode) is necessary but not sufficient: even with `AiDotNet.Tensors` v0.80.2 + an eager-path control proving the gradient pipeline is correct end-to-end, the fused path still loses dL/dW everywhere except the last few additive params.

## Repro

Master + the existing `TransformerByteLMDiagnostic1328Tests` test suite + Tensors v0.80.2. With `TrainingDiagnosticsConfig.ForceEagerPath = false` (default), run a one-step `Train()` on the 2-encoder, 4-head, d_model=32 Transformer used in #1328 diagnostics, then enumerate every trainable parameter via `model.Layers[i].GetTrainableParameters()` and compare each tensor's L2 norm before vs after.

## Observation matrix (29 params, single Train() call, lr=0.001, Adam)

```
FUSED PATH                              EAGER PATH (control, ForceEagerPath=true)
================                        =========================================
[STUCK] layer 00 EmbeddingLayer    [8,32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 03 MHA Wq           [32,32]  [MOVED] L2(∆)=3.16e-2
[STUCK] layer 03 MHA Wk           [32,32]  [MOVED] L2(∆)=2.72e-2
[STUCK] layer 03 MHA Wv           [32,32]  [MOVED] L2(∆)=3.20e-2
[STUCK] layer 03 MHA Wo           [32,32]  [MOVED] L2(∆)=3.20e-2
[STUCK] layer 03 MHA bias            [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 04 LayerNorm γ         [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 04 LayerNorm β         [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 06 Dense W          [32,64]  [MOVED] L2(∆)=3.59e-2
[STUCK] layer 06 Dense b             [64]  [MOVED] L2(∆)=6.48e-3
[STUCK] layer 07 Dense W          [64,32]  [MOVED] L2(∆)=3.67e-2
[STUCK] layer 07 Dense b             [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 08 LayerNorm γ         [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 08 LayerNorm β         [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 10 MHA Wq           [32,32]  [MOVED] L2(∆)=2.94e-2
[STUCK] layer 10 MHA Wk           [32,32]  [MOVED] L2(∆)=3.20e-2
[STUCK] layer 10 MHA Wv           [32,32]  [MOVED] L2(∆)=3.20e-2
[STUCK] layer 10 MHA Wo           [32,32]  [MOVED] L2(∆)=3.20e-2
[STUCK] layer 10 MHA bias            [32]  [MOVED] L2(∆)=5.66e-3
[STUCK] layer 11 LayerNorm γ         [32]  [MOVED] L2(∆)=5.39e-3
[STUCK] layer 11 LayerNorm β         [32]  [MOVED] L2(∆)=5.39e-3
[STUCK] layer 13 Dense W          [32,64]  [MOVED] L2(∆)=3.28e-2
[STUCK] layer 13 Dense b             [64]  [MOVED] L2(∆)=6.08e-3
[STUCK] layer 14 Dense W          [64,32]  [MOVED] L2(∆)=3.44e-2
[MOVED] layer 14 Dense b      L2=5.66e-3   [MOVED] L2(∆)=5.66e-3
[STUCK] layer 15 LayerNorm γ         [32]  [MOVED] L2(∆)=5.29e-3
[MOVED] layer 15 LayerNorm β  L2=5.20e-3   [MOVED] L2(∆)=5.29e-3
[STUCK] layer 18 Dense W          [32,8]   [MOVED] L2(∆)=1.50e-2
[MOVED] layer 18 Dense b      L2=2.83e-3   [MOVED] L2(∆)=2.83e-3
```

**Summary:** fused = 3 moved / 26 stuck. eager = 29 moved / 0 stuck.

## Pattern

The 3 params that survive in the fused path are all 1D additive params (bias / β) on the layers nearest the output. Every 2D weight matrix is stuck, including ones (e.g. layer 06 Dense W) whose same-layer bias also stuck. This is **not** an embedding-specific bug — it's a broader gradient-binding failure in the compiled-plan backward chain.

## Hypothesis

`AiDotNet.Tensors.CompiledTrainingPlan` either:
- (a) records the lazy graph correctly but does NOT compile a backward node for every reachable parameter — only a subset survives the trace → compile → replay pipeline, or
- (b) compiles backward nodes for all params but the fused-Adam in-place update kernel doesn't bind gradient buffers to all parameter slots, or
- (c) has an op-specific backward gap (e.g., MatMul backward `dW = X^T @ dY` writes only when X is a leaf input, not when X is intermediate), so weight grads die wherever their MatMul input is an intermediate tensor (which is almost everywhere in a multi-layer network).

The bias survival pattern points at (c) most strongly: biases get `db = sum(dY)` which only needs the gradient flowing into the op, not the saved-state input. Weights need `X^T @ dY` which needs the saved input as well — if that saved input is dropped for non-leaf X, dW = 0.

## Eager-path control

`Diagnostic_EagerPath_OnlyConvergeCheck` (with `ForceEagerPath=true`) achieves `avgNll = 0.0000, top-1 = 8/8` at epoch 500 under v0.80.2. The eager path's `TrainWithTape` walks the full backward correctly. Confirms the bug is fused-path-specific.

## Current workaround

`AiDotNet.Configuration.TrainingDiagnosticsConfig.ForceEagerPath = true` (shipped in #1330) routes around the fused path entirely. Trains correctly at ~10–50% the speed of a working fused kernel.

## Why this matters

Every multi-layer model using AdamOptimizer/AdamWOptimizer/SGDOptimizer with the default `EnableCompilation = true` will silently degrade to "biases learn, weights don't" — visually it looks like the model is training (loss decreases slightly), but the network is effectively frozen at random init. #1328 was caught because byte-LM degenerates to `PPL = V` exactly when only biases move; subtler regressions (e.g., image classifiers) might just appear to learn slowly.

## Next steps

1. Add per-parameter delta probe to `AiDotNet.Tensors.Tests` directly (without going through AiDotNet's `Transformer`) so the bug is reproduced inside the Tensors repo.
2. Instrument `CompiledTrainingPlan.Step` to log the gradient buffer for each registered parameter — distinguish "buffer is zero" from "buffer is not bound to param".
3. Audit each op's backward registration under `GraphMode.RecordBinary` / `RecordVariadic` for whether the W-side gradient is propagated to leaf parameters.

cc @ooples — this blocks any production training that uses the fused path.

**Comment by ooples** [comment]
## Investigation update — Tensors-only repro is NEGATIVE

Pushed a Tensors-side integration suite isolating the same op patterns: https://github.com/ooples/AiDotNet.Tensors/tree/fix/1331-fused-adam-multilayer-investigation (file `tests/AiDotNet.Tensors.Tests/Engines/Compilation/MultiMatMulFusedAdamParamUpdateTests.cs`). Every test asserts L2(∆) > 0 on every leaf parameter after one Adam Step.

**All 5 cases PASS** on net471 and net10.0 — the Tensors compiled-plan + fused-Adam path works correctly for the primitives AiDotNet uses:

| Test | Result |
|---|---|
| Two stacked `MatMul + BroadcastAdd` | ✅ all 4 params move |
| Three stacked `MatMul + BroadcastAdd` | ✅ all 6 params move |
| Two stacked `FusedLinear` (no activation) | ✅ all 4 params move |
| Two `FusedLinear` with rank-3 ↔ rank-2 Reshape | ✅ all 4 params move |
| `EmbeddingLookup → Reshape → FusedLinear → Reshape → FusedLinear → Reshape` | ✅ all 5 params (incl. embedding) move |

So the gradient-binding contract (`gradMap[parameters[i]] → dL/dparam → fused-Adam in-place update`) is intact for every op the AiDotNet `DenseLayer.Forward` + `EmbeddingLayer.Forward` paths emit, in the order they emit them, with the AiDotNet shape conventions.

## Implication: the bug is NOT in Tensors primitives

The Transformer-level 26/29-stuck-params fingerprint must come from one of:

1. **An op the test suite doesn't cover yet** — `LayerNormalizationLayer` and `MultiHeadAttentionLayer` both stuck completely (4 of 4 LN pairs, 0 of 5 MHA matrices moved in the original census). These are the prime suspects. Their GraphMode recording or backward graph may drop the input chain.
2. **`Predict()` → `Train()` interaction** — the AiDotNet diagnostic calls `model.Predict()` before `model.Train()`. The eager Predict() path through `EmbeddingLayer.Forward` calls `DifferentiableOps.RecordIfActive(...)` even without a tape, possibly setting per-tensor state (`_gradIndex`, `_pinnedByTape`) that the subsequent GraphMode trace then misinterprets.
3. **`_cachedParameters` ref drift** — `CompiledTapeTrainingStep.TryStepWithFusedOptimizer` caches `parameters` from the first `forward(input)` call. If any layer's Forward replaces a parameter Tensor<T> reference on a subsequent forward (e.g. via `AllocateLazyWeight` returning a pooled buffer with different ref identity), the cached array points to a stale tensor while the trace records against the new ref → `gradMap.ContainsKey(parameters[i])` returns false → `gradients[i]` is null → fused-Adam skips it.

## Next steps (in priority order)

- [ ] Add a Tensors test that includes `LayerNorm` (which DenseLayer-FFN feeds into) and verify all gain/beta + upstream W params move.
- [ ] Add a Tensors test that mimics the Transformer's MHA shape: 4× linear projections (Q, K, V, O) sharing the same input. The 5 MHA params (Wq, Wk, Wv, Wo, bias) were all stuck in the Transformer; verify whether 4 stacked FusedLinears feeding the same softmax-attention block reproduce that.
- [ ] In the AiDotNet probe, instrument `CompiledTapeTrainingStep.TryStepWithFusedOptimizer` to print: count of parameters, count of gradients in `plan.Gradients` that are non-null, and the L2 of each. This pinpoints whether the bug is "gradient not computed" vs "gradient computed but bound to wrong tensor ref".
- [ ] Verify that `_cachedParameters` doesn't go stale by comparing `_cachedParameters[i]` references against `model.Layers[*].GetTrainableParameters()` after a Predict + first Train.

The Tensors-side regression suite stays valuable regardless — it pins the existing contract so future Tensors work can't silently regress these chains.

**Comment by ooples** [comment]
## Scientific narrow-down — root cause identified at the buffer-aliasing layer

Pushed comprehensive Tensors-side fixes on `fix/1331-fused-adam-multilayer-investigation` (PR #356):
- LayerNorm compile-time mean/variance staleness — **fixed**, refresh on every plan.Step
- Same compile-time savedState refresh applied to BatchNorm, GroupNorm, InstanceNorm, RMSNorm, Dropout, ScatterMean
- GPU engine GraphMode bypass — **fixed**, delegates to base CpuEngine when GraphMode active (LayerNorm, RMSNorm, GroupNorm, InstanceNorm, Dropout)
- Dropped broken `OpType.LayerNorm` specialized forward that overwrote the lazy execute lambda

These are real, isolated, regression-tested fixes (8/8 Tensors-side tests pass).

## End-to-end Transformer still shows 26/29 stuck — actual root cause

Trace instrumentation pinpoints the deeper issue. With every fix above applied and `AiDotNetEngine.Current` pinned to `CpuEngine`, the LayerNorm's lazy execute callback runs and observes:

```
[LN-LAZY-EXEC] ci.L2=0.0000E+000 ci.Shape=[1,4,32] capturedInputRef==ci:True ci.LazySource==null:False
```

**Translation:** at the moment LayerNorm's compiled forward step runs (after the upstream compiled forward steps have supposedly written), `ci.GetFlattenedData()` returns all-zero data. The reference is the same as recording time (`capturedInputRef==ci:True`), the LazySource pointer is still attached, but the backing array is zero.

This means **the previous op's compiled forward writes its output to a different backing array than what the next op reads from**. In other words: the lazy-graph compile-time buffer that LayerNorm captured as `ci` is NOT the same backing array that the upstream FusedLinear / Permute / Reshape / etc. writes to at plan.Step forward replay.

Possible mechanisms:
1. `TryBuildSpecializedForward` captures `step.OutputBuffer.GetDataArray()` at compile time and writes there, but if `step.OutputBuffer` is later replaced (e.g. by a `.Contiguous()` materialisation or pool-rent) the next op's `step.Inputs[i]` still points at the OLD ref → reads zero.
2. The compile-time topology has buffer-sharing between unrelated ops via the tensor pool; one op's RentUninitialized happens to return the same array slot that another op's RentUninitialized later overwrites.
3. The graph compiler reorders forward steps in a way that breaks the data-dependency the producer/consumer pair was recorded with.

## What's NOT the cause (eliminated by the trace)

- ✗ Not a parameter-reference mismatch — `capturedInputRef==ci:True` confirms identity preserved
- ✗ Not a GraphMode bypass — `GraphMode.IsActive=True` at the record site for all 4 LayerNorm calls
- ✗ Not an engine-dispatch issue — `engineType=CpuEngine` confirmed
- ✗ Not a missing backward function — `BackwardFn` set, savedState delivered
- ✗ Not a specialized fast-path overwrite — that's been disabled for LayerNorm
- ✗ Not a compile-time-stale mean/variance — `freshMean.L2` would be non-zero IF `ci.L2` weren't zero; but ci is zero so fresh mean comes out zero correctly

## Forward-step dump showed the topology

```
[FWDSTEP 0] TensorEmbeddingLookup [4,32]
[FWDSTEP 1] TensorPermute [1,4,4,8]
[FWDSTEP 2] TensorPermute [1,4,4,8]
...
[FWDSTEP 11] LayerNorm [1,4,32]
[FWDSTEP 12] LayerNorm [1,4,32]
...
```

Multiple duplicate-shape ops at adjacent indices (likely Q/K projections, multiple LayerNorms across encoders) — combined with `TensorMatMul` appearing AT step 38 AFTER `Softmax` (step 37) and `TensorAddScalar` (step 39), the topology is unusual enough to warrant investigation of LazyGraphCompiler's ordering and DCE passes.

## Next investigation step

Need to instrument `step.OutputBuffer.GetDataArray()` for each compiled forward step AND `step.Inputs[i].GetDataArray()` for the next consumer, comparing `ReferenceEquals` on the underlying arrays. If they differ, the buffer-aliasing in the lazy-graph compiler is the bug.

This is genuinely a deeper architectural issue in the lazy graph + compiled-plan buffer management, separate from the LayerNorm savedState issue that PR #356 fixes.

**Comment by ooples** [comment]
Closing — this was already fixed by PR #1330 (merged 2026-05-16, commit 48f79fc2). The PR body said "Closes #1328 and #1331" but GitHub's auto-close didn't fire on that exact phrasing.

The fix has two parts:

1. **AiDotNet.Tensors side** (PR #356, shipped in Tensors 0.80.3, now 0.81.3):
   - `CompiledModelCache.GetOrCompileTraining` now refreshes input/target data on cache hits via `_persistentInput`/`_persistentTarget` in `CompiledTapeTrainingStep`
   - `TensorEmbeddingLookupFromFloatIndices` captures float-indices tensor by reference (vs the old snapshot-at-trace-time behavior)
   - `ScaledDotProductAttention` GraphMode branch decomposed into tape-recorded primitives instead of wrapping a recursive eager call
   - LayerNorm / RMSNorm / BatchNorm / GroupNorm / InstanceNorm / Dropout / ScatterMean savedState refreshed on every replay

2. **AiDotNet side** (PR #1330): consumed the Tensors fix + added regression tests `Transformer_ByteLM_FusedPath_AllParamsMustMove` and `Transformer_ByteLM_FusedPath_ConvergesAfter1331Fix` (500 epochs, fused path, avgNll 10.77 → 0.10, top-1 = 8/8).

Verified-by-test on master + Tensors 0.81.3: all 29 trainable params in the 2-encoder Transformer encoder now move after a single fused `Train()` step (vs the original 3-moved/26-stuck observation matrix in this issue).
